### ISSUE TITLE
Bloomberg feedback on TypeScript 4.5 Beta

### ISSUE THREAD

**Issue by mkubilayk** [microsoft/TypeScript#46382]
_This issue has been created by the team responsible for the infrastructure behind most of Bloomberg's TypeScript code._

We have tested TypeScript 4.5 Beta on our TypeScript codebase. The current results indicate, once again, a fairly easy upgrade.

- We did not need any change in our integration code using TypeScript compiler APIs.
- JavaScript output slightly changed (`target: "esnext"`). These are explained in detail below.
- Declaration output slightly changed. These are explained in detail below.
- Some of our TypeScript code required changes to make it compile with TypeScript 4.5 Beta. These are explained in detail below.
- We are still far from being able to use `importsNotUsedAsValues: "error"`, `preserveValueImports` and `isolatedModules` at the same time.

## Impact: Emit

| # | Change | Affects | Release notes | Packages affected | Introduced in |
|--:|:--|:--|:--|:--|:--|
| 1 | JSX transform spreads children | JS Emit | Not announced | <1% | #45693 |
| 2 | `Promise.all` overload for `any` | DTS Emit | [Announced][promise-improvements] | <1% | #45350 |

### 1. JSX transform spreads children

This is the only change in JavaScript output in our codebase. It is not a breaking change in semantics as `React.createElement` can handle both cases.

```ts
// TS input
() => <>{...[]}</>;
```

```ts
// JS output in TS 4.5 Beta
() => React.createElement(React.Fragment, null, ...[]);
```

[Playground in TS 4.5 Beta](https://www.typescriptlang.org/play?ts=4.5.0-beta#code/BQSgBAvAfGA8UG8B0KDaBdAvrA9FIA)

```ts
// JS output in TS 4.4.3
() => React.createElement(React.Fragment, null, []);
```

[Playground in TS 4.4.3](https://www.typescriptlang.org/play?ts=4.4.3#code/BQSgBAvAfGA8UG8B0KDaBdAvrA9FIA)

### 2. [`Promise.all` overload for `any`][promise-improvements]

This is the only change in declaration output in our codebase. It is not explicitly mentioned in the release notes. It might be possible to handcraft some examples where this could be a breaking change but that is not the case for our codebase. This change looks harmless.

```ts
// TS input
function f() { return Promise.all([] as any); }
```

```ts
// DTS output in TS 4.5 Beta
declare function f(): Promise<{
    [x: string]: any;
}>;
```

[Playground in TS 4.5 Beta](https://www.typescriptlang.org/play?ts=4.5.0-beta#code/GYVwdgxgLglg9mABMAFASkQb0QJwKZQg5IAKOcAtjAM54B0AhgDZMoDaAuog9d2AJ5oA3IgC+QA)

```ts
// DTS output in TS 4.4.3
declare function f(): Promise<[unknown, unknown, unknown, unknown, unknown, unknown, unknown, unknown, unknown, unknown]>;
```

[Playground in TS 4.4.3](https://www.typescriptlang.org/play?ts=4.4.3#code/GYVwdgxgLglg9mABMAFASkQb0QJwKZQg5IAKOcAtjAM54B0AhgDZMoDaAuog9d2AJ5oA3IgC+QA)

## Impact: Type Checking

| #   | Change                                          | Release notes                     | Packages affected | Introduced in                         |
| --- | ----------------------------------------------- | --------------------------------- | ----------------- | ------------------------------------- |
| 1   | Awaited type & promise signature change         | [Announced][promise-improvements] | <1%               | #45350                                |
| 2   | Change of inference with `noImplicitAny: false` | Not announced                     | <1%               | #45350                                |
| 3   | Edge case with generics and indexing            | Not announced                     | <1%               | #41821                                |

The overall impact of this version is very small and we can proactively update all affected packages in a forward-compatible way. However, we do think that some corner cases are worth looking at as they break here.

### 1. [Awaited type & promise signature change][promise-improvements]

This is a documented change that seems to be breaking in very small corner cases.

Here is a first example where `Awaited` trips up with control flow analysis in TS 4.5 Beta:

```ts
declare function deepEqual<T>(actual: T, expected: T): void;

export async function testRequestSuccess<T>(resultP: Promise<T>, expected: T) {
    let actual;
    try {
        actual = await resultP;
    } catch (_e) {}

    // The type parameter of `deepEqual` is inferred based on the first argument.
    // TypeScript 4.4 calculates `T | undefined`, so passing `expected: T` is okay.
    // TypeScript 4.5 calculates `Awaited<T> | undefined`, which can't have `T` assigned to it.
    // Removing the earlier try-catch gets things working because `actual`'s type is just `Awaited<T>`,
    // which you *can* assign `T` to.
    deepEqual(actual, expected);
}
```

[Playground in TS 4.4.3](https://www.typescriptlang.org/play?ts=4.4.3#code/CYUwxgNghgTiAEAzArgOzAFwJYHtX1BAAcBRAR2SggB4AVAPgAoAoeN+KTSiALnloA0rdiAAeRcBhDA+g5gEo+ANxxZgAbmbMxRHDAwcAzgE90SNJlz4phjACUQFELYDKyMGGeG6TYWziGyBAYAAp8ITA4ALZYhiA+QuzwOpLSskLy8ADefvAQIAacGNyaSRgwxtm5SUXc8AC8HADuUFgGAUGhpewAvvBgUBhgABbwjAD6IJk5ST1aSQD0C-zDCBjGEvBEsFBRBSAw8DiI8AAGhKQUVKfwsbeoiAdwwPAARlBxL3jwGKtIWDBbBwYABzZB7VAYAB0uSW-A2IBcYBgWCIBgALFD0f0qGAgoNnGdaPAAD7wNCgRBYVDSU4CeCGHBbD6Gakgs4pTBpfg3O44ADWUGMMMWy1oCKRKLR8ExAFYcRA8dAbGcAIItNrSHyk8moSnU2n0prDLAjHGoADkBmGUCUCFOtBuLKwIJpLwwTLaIvYcIcURwSjZPz+IFgECwBx+FQAtAMhqMQQVDMG2cmmnp+UHXuAoMg4mdatcLcn1ps7gArPMGU7q1pSYA+Omw5bG02jYw4ZDwABUA1Q3aMrNdRJuHu9bAu5G4jELEHpnPr8k0PSAA) | [Playground in TS 4.5 Beta](https://www.typescriptlang.org/play?ts=4.5.0-beta#code/CYUwxgNghgTiAEAzArgOzAFwJYHtX1BAAcBRAR2SggB4AVAPgAoAoeN+KTSiALnloA0rdiAAeRcBhDA+g5gEo+ANxxZgAbmbMxRHDAwcAzgE90SNJlz4phjACUQFELYDKyMGGeG6TYWziGyBAYAAp8ITA4ALZYhiA+QuzwOpLSskLy8ADefvAQIAacGNyaSRgwxtm5SUXc8AC8HADuUFgGAUGhpewAvvBgUBhgABbwjAD6IJk5ST1aSQD0C-zDCBjGEvBEsFBRBSAw8DiI8AAGhKQUVKfwsbeoiAdwwPAARlBxL3jwGKtIWDBbBwYABzZB7VAYAB0uSW-A2IBcYBgWCIBgALFD0f0qGAgoNnGdaPAAD7wNCgRBYVDSU4CeCGHBbD6Gakgs4pTBpfg3O44ADWUGMMMWy1oCKRKLR8ExAFYcRA8dAbGcAIItNrSHyk8moSnU2n0prDLAjHGoADkBmGUCUCFOtBuLKwIJpLwwTLaIvYcIcURwSjZPz+IFgECwBx+FQAtAMhqMQQVDMG2cmmnp+UHXuAoMg4mdatcLcn1ps7gArPMGU7q1pSYA+Omw5bG02jYw4ZDwABUA1Q3aMrNdRJuHu9bAu5G4jELEHpnPr8k0PSAA)

Additionally, the change in type signature of `Promise` has brought up another issue with an explicit `Promise.all` typing:

```ts
// This call breaks in TS 4.5 Beta because
// multiple type variables are no longer supported
Promise.all<string, number>(["a", 1]);
```

[Playground in TS 4.4.3](https://www.typescriptlang.org/play?ts=4.4.3#code/PTAEBUAsEsGdQMYEMA2LQCMBOBTJBreaAOwgGVQAWAOgFZMdkBXWHUAWyZQBdoAHFG24BPPmwBuSLNCQZB8KW2IB7UCmXEA5jiyhYTPn2VZuOACYAuAFAAFLMvZwc1VCgA8sbtK0AaUMSZ2DB0APgAKAG0AIiQovwBGAF0ASiA) | [Playground in TS 4.5 Beta](https://www.typescriptlang.org/play?ts=4.5.0-beta#code/PTAEBUAsEsGdQMYEMA2LQCMBOBTJBreaAOwgGVQAWAOgFZMdkBXWHUAWyZQBdoAHFG24BPPmwBuSLNCQZB8KW2IB7UCmXEA5jiyhYTPn2VZuOACYAuAFAAFLMvZwc1VCgA8sbtK0AaUMSZ2DB0APgAKAG0AIiQovwBGAF0ASiA)

Given the low prevalence of those use cases, we can easily convert them to the new form:

```ts
Promise.all<[string, number]>(["a", 1]);
```

[Playground in TS 4.5 Beta](https://www.typescriptlang.org/play?ts=4.5.0-beta#code/AoJw9gtglgzgpgOgIYBsUB4DaMAuIoB2A5gDQAEBArhAEZwgC6AfABSYBES75AjAwJRA)

### 2. Change of inference with `noImplicitAny: false`

Some of our projects have the `noImplicitAny: false` flag set. In this situation, some unwanted inference seems to happen in TS 4.5 Beta:

```ts
interface RejectedValue {
    status: "rejected";
    reason: Error | any;
}
interface FulfilledValue {
    status: "fulfilled";
    value: unknown;
}
type SettledValue = RejectedValue | FulfilledValue;
type SettledPromises = Promise<SettledValue[]>;

// This example requires `"noImplicitAny": false`.
export default function allSettled(promises): SettledPromises {
    const wrappedPromises = promises.map((p) =>
        Promise.resolve(p)
            .then((value) => ({ status: "fulfilled", value }))
            .catch((reason) => ({ status: "rejected", reason }))
    );
    return Promise.all(wrappedPromises);
//  ^ TS2322
//    Type 'Promise<{ [x: string]: any; }>' is not assignable to type 'SettledPromises'.
}
```

[Playground in TS 4.4.3](https://www.typescriptlang.org/play?noImplicitAny=false&ts=4.4.3#code/JYOwLgpgTgZghgYwgAgEoQFYQZAJgNTgBsBXFAbwChkbkBnMOMEugLmQCIpNs8OBuarW5w6AexDsAolChioyAD7I4IAJ6CAvpVCRYiFADESRGMCJEIBYmWRVa9RszacYJsxasChNAG42IdhIQAGsQMQB3EC1KMDUABxQAZQgwMEtrUhQAXjQeHCtCLKVkY1NzDKKyQTjE5BS0jIAFOQBbYDoIOmRclrF2zoAeBvTCgIBtAF0APkFKAHp55AAVAAsO5AgADzhW+MtkbgBHEmBuboADDnCAST2iYARgMABBdQ52eCJOi4A6Sm28XkYGQuAg8BMILcIBwwAkKgsIwyAAp4m0Ol0AJTsJFWPoDLp2HzIBASBjICJQODxRK4fEY7q5NH9Bm-VrU5GozE9abEhz0zq-c5iIi+CBcvkOWi-MCrCAgTn+LLc7LTZDI8iOJgsdgcNzlTy4DgAGmQStsmkxmMlUuQvwQTAQq05InEIBVao1Wucuu4WAKRtNrvhlutDkxggc3GYUBAyAFEF+xCIyMp1NpCboEcomiAA) | [Playground in TS 4.5 Beta](https://www.typescriptlang.org/play?noImplicitAny=false&ts=4.5.0-beta#code/JYOwLgpgTgZghgYwgAgEoQFYQZAJgNTgBsBXFAbwChkbkBnMOMEugLmQCIpNs8OBuarW5w6AexDsAolChioyAD7I4IAJ6CAvpVCRYiFADESRGMCJEIBYmWRVa9RszacYJsxasChNAG42IdhIQAGsQMQB3EC1KMDUABxQAZQgwMEtrUhQAXjQeHCtCLKVkY1NzDKKyQTjE5BS0jIAFOQBbYDoIOmRclrF2zoAeBvTCgIBtAF0APkFKAHp55AAVAAsO5AgADzhW+MtkbgBHEmBuboADDnCAST2iYARgMABBdQ52eCJOi4A6Sm28XkYGQuAg8BMILcIBwwAkKgsIwyAAp4m0Ol0AJTsJFWPoDLp2HzIBASBjICJQODxRK4fEY7q5NH9Bm-VrU5GozE9abEhz0zq-c5iIi+CBcvkOWi-MCrCAgTn+LLc7LTZDI8iOJgsdgcNzlTy4DgAGmQStsmkxmMlUuQvwQTAQq05InEIBVao1Wucuu4WAKRtNrvhlutDkxggc3GYUBAyAFEF+xCIyMp1NpCboEcomiAA)

In this example, return value is inferred to have `Promise<{ [x: string]: any; }>` type. This conflicts with the type annotation.

### 3. Edge case with generics and indexing

We found an instance where behavior differs from the previous version. The following example is contrived but is extracted from a much larger use case in our codebase.

```ts
type TypesOf<T extends Object> = T[keyof T]
type MatchingKeys<T extends Record<string,any>,  Condition>
    = TypesOf<{ [K in keyof T ]: T[K] extends Condition ? K : never }>


export interface Model {
    foo: string;
    bar: Array<string>;
}

type Extracted<A extends Array<any>> = unknown;
export declare function extract<
    M extends Model,
    K extends MatchingKeys<M, Array<any>>
>(m: M): Extracted<M[K]>;
//                 ^ TS2344
//                   Type 'M[string]' is not assignable to type 'any[]'.
```

[Playground in TS 4.4.3](https://www.typescriptlang.org/play?ts=4.4.3#code/C4TwDgpgBAKuEGcDyAzAPDKEAewIDsATBKJAIwCsIBjYAPigF5YBtAawhAHsVYBdAFChIUALIBDYNQAWAS3wBzANKcEGLLgLEoAJRpcAToTQJgB+QoA04-CDqWoUAMJcis4LNcMBjx8ziQyOgA3lAsSlDyUBzcvJh8AFysSnwaeEQkLm4erlAA-FARSfgQAG4QBlAAvnQCdThghsCR+HgGKOLU0KJchBAANlDBPr4oXFxJpuaKANwjjmTiBkkAggYG4iAmZhZ0c1V1QvBQAKK4G7QQxitpWiRrG1s2dgzMAK74bPhcAO74cw0mlA+tR+ktoCgPrRPPg0hdgGhRLcMmJegMHBEcOltBIpHJFCoQGpRA4Hps0M86FSABQAWySogAlEkzmZOnhjKJwnw9gIgA) | [Playground in TS 4.5 Beta](https://www.typescriptlang.org/play?ts=4.5.0-beta#code/C4TwDgpgBAKuEGcDyAzAPDKEAewIDsATBKJAIwCsIBjYAPigF5YBtAawhAHsVYBdAFChIUALIBDYNQAWAS3wBzANKcEGLLgLEoAJRpcAToTQJgB+QoA04-CDqWoUAMJcis4LNcMBjx8ziQyOgA3lAsSlDyUBzcvJh8AFysSnwaeEQkLm4erlAA-FARSfgQAG4QBlAAvnQCdThghsCR+HgGKOLU0KJchBAANlDBPr4oXFxJpuaKANwjjmTiBkkAggYG4iAmZhZ0c1V1QvBQAKK4G7QQxitpWiRrG1s2dgzMAK74bPhcAO74cw0mlA+tR+ktoCgPrRPPg0hdgGhRLcMmJegMHBEcOltBIpHJFCoQGpRA4Hps0M86FSABQAWySogAlEkzmZOnhjKJwnw9gIgA)

We will attempt to simplify it in the larger case to avoid this situation. However, this is still a breaking change for this edge case.

## Additional Remarks: `preserveValueImports`

We are attempting to enable a few compiler options around modules that simplify the TS -> JS transform.

In order to do so we attempted to set `importsNotUsedAsValues: "error"` and `preserveValueImports`.  We already have `isolatedModules` fully enabled and `importsNotUsedAsValues: "error"` partially enabled.

40% of our packages are ready for those options as-is while the remaining 60% are at least getting one `TS1444` error or more.  This is to be expected as many users import a mix of values and types on a single line.

[`type` Modifiers on Import Names][type-modifiers] will be helpful to resolve those errors and promote those options further.

A code fix that is capable of inserting `type` modifiers where appropriate would be very valuable here. It would turn the following code:

```ts
import { value, Type } from "./lib";
//              ^ TS1444
// 'Type' is a type and must be imported using a type-only import when 'preserveValueImports' and 'isolatedModules' are both enabled.
```

into

```ts
import { value, type Type } from "./lib";
// no errors 🎉
```

---

If any of these changes do not look intentional/right and should be considered as a bug, we can open issues.

_Credits to @scottrobertwhittaker, @rricard, @tchetwin and @robpalme for helping with this write up._

[promise-improvements]: https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#the-awaited-type-and-promise-improvements
[type-modifiers]: https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#type-on-import-names

**Comment by andrewbranch** [comment]
Thanks as always for the detailed feedback! I can chime in on a couple things:

I think “`Promise.all` overload for `any`” and “Change of inference with `noImplicitAny: false`” are in fact the same issue, which is being tracked at https://github.com/microsoft/TypeScript/issues/46169.

> A code fix that is capable of inserting `type` modifiers where appropriate would be very valuable here.

I intended to add this codefix when I implemented type-only import specifiers, but must have forgotten. I think we can probably get this in before the RC.
